### DISCUSSION TITLE
How to have dynamic "where" in query builder ?

### DISCUSSION THREAD

**Discussion by Barbapapazes** [adonisjs/core#2530] [Help] [answered]
Hello,

In the documentation, we can see this: 
```js
const users = await User
  .query() // 👈now have access to all query builder methods
  .where('country_code', 'IN')
  .orWhereNull('country_code')
```

But, if my client application have many filters, how to attach, depending on how many filters the client use, multiple "where" to the query ? Sometime, 1 filter, and sometime 4. How to manage this ?

 ```js
const users = await User
  .query() // 👈now have access to all query builder methods
  .where('country_code', 'IN')
  .where('another_country_code', 'IN')
  .where('another_filter', 'IN')
  .orWhereNull('country_code')
```

Thanks a lot fo answer !

**Comment by Barbapapazes** [comment]
No answer needed ! I have found a solution and will post it here soon !

**Comment by Vong3432** [comment]
I have tried to add one for this, but not sure is it good tho. 

### ORMService.ts
<img width="518" alt="image" src="https://user-images.githubusercontent.com/36991424/129444959-982960ed-7481-49c0-9789-169579967846.png">
<img width="604" alt="image" src="https://user-images.githubusercontent.com/36991424/129444964-4b338497-ec31-45e4-a1f2-b03757a262fc.png">

### Then in Controller/Service file
```
const results = await new ORMFilterService(ModelA, request.qs()).getFilteredResults()
return results as ModelA[];
```

### Explanation
What I did is I try to put the `ModelA` and `request.qs()` in the `ORMFilterService(ModelA, request.qs())` class,
when `getFilteredResults()` is called, it loops through the `request.qs()` and insert the `keys` and `value` to `keyArrays` and `valArrays`. Finally, use the `whereIn` method to do the filter (The value has to be the same as the columns in table).

**Comment by binsarjr** [comment]
Just do it like this
```javascript
const {
    country_code: countryCode,
    another
} = request.all() // data sample

const query = User.query()
if(countryCode) query.where('country_code', 'IN')
if(another) {
  query.where('another_country_code', 'IN')
  .where('another_filter', 'IN')
}

query.orWhereNull('country_code')
const users = await query.finally()
```

I hope this helps

**Comment by RomainLanz** [comment, marked as answer]
Hey @Barbapapazes! 👋 

You can use the `if` method from the QueryBuilder.

```ts
User
  .query()
  .if(searchQuery, (query) => {
    query.where('first_name', 'like', `%${searchQuery}%`)
    query.where('last_name', 'like', `%${searchQuery}%`)
  })
```

📚 https://docs.adonisjs.com/reference/database/query-builder#if

You can use an `if` statement like you would do.

```ts
const query = User.query()

if (foo) {
  query.where(...)
}

if (bar) {
  query.where(...)
}
````

**Reply by Barbapapazes** [reply]
Yeap, thanks a lot! 👌
